perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java

```diff
@@ -2971,19 +3544,26 @@ public void deleteAllAttributeAuthz(PerunSession sess, AttributeDefinition attri
 
 	@Override
 	public List<Attribute> getRequiredAttributes(PerunSession sess, Resource resourceToGetServicesFrom, Host host) throws InternalErrorException {
+		List<Attribute> attributesCache = null;
+
 		if(!CacheManager.isCacheDisabled() && !perun.getCacheManager().wasCacheUpdatedInTransaction()) {
 			List<Integer> attrIds = getRequiredAttributeIds(resourceToGetServicesFrom);
 			List<Attribute> attrs = perun.getCacheManager().getAttributesByIds(attrIds, new Holder(host.getId(), Holder.HolderType.HOST));
-			return this.setValuesOfAttributes(sess, attrs, host, null);
+			attributesCache = this.setValuesOfAttributes(sess, attrs, host, null);
 		}
 
 		try {
-			return jdbc.query("select " + getAttributeMappingSelectQuery("host") + " from attr_names " +
+			List<Attribute> attributesDB = jdbc.query("select " + getAttributeMappingSelectQuery("host") + " from attr_names " +
 							"left join      host_attr_values   host    on      attr_names.id=host.attr_id    and   host_id=? " +
 							"where namespace in (?,?,?) " +
 							"and attr_names.id in (select distinct service_required_attrs.attr_id from service_required_attrs " +
 							"join resource_services on service_required_attrs.service_id=resource_services.service_id and resource_services.resource_id=?)",
 					new SingleBeanAttributeRowMapper<>(sess, this, host), host.getId(), AttributesManager.NS_HOST_ATTR_CORE, AttributesManager.NS_HOST_ATTR_DEF, AttributesManager.NS_HOST_ATTR_OPT, resourceToGetServicesFrom.getId());
+
+			String message = "The attributes do not match, attributes from cache are: " + attributesCache + " while attributes from db are: " + attributesDB + " the attributes belong to " + host + " from service with resource: " + resourceToGetServicesFrom;
```

Comment:
The resource only specify, which services required attributes will be retrieved for the host. I'm not sure we need to log that. It somehow complicate my idea about generic message in private method with passed params, since you need to know their number. You can probably check its length and create placeholders in the message accordingly.

When I think about it, it would be more interesting to print stack trace, to identify which method failed on this check. I'm not sure, if there is any other way, than throwing and catching exception inside the private method to get stack trace.